### PR TITLE
exp/ingest: Add Reporter interface for sessions

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### Summary

A Reporter is used by a session to report progress while executing state / ledger pipelines

Fixes https://github.com/stellar/go/issues/1459

### Goal and scope

We want to provide a mechanism to log progress and/or record metrics as a session processes its pipelines.

### Summary of changes

Define a reporter interface which is invoked by sessions during significant events such as:
the session as started to process a new history archive snapshot / ledger
the session has processed a new entry from the StateReader / LedgerReader
the session has finished processing the history archive snapshot / ledger

### What shouldn't be reviewed

N / A

### CODE REVIEW

File: exp/ingest/main.go [hunk 1]
```diff
@@ -77,3 +82,71 @@ type Session interface {
...
+// reporterLedgerReader instruments a io.StateReader with a StateReporter
+// which reports each xdr.LedgerEntryChange which is read from the reader
+type reporterStateReader struct {
+	io.StateReader
+	StateReporter
+}
+
+func (r reporterStateReader) Read() (xdr.LedgerEntryChange, error) {
+	entry, err := r.StateReader.Read()
+	if err != nil {
```
Comment (bartekn) [inline exp/ingest/main.go]:
```suggestion
	if err == nil {
```
Currently it calls `OnStateEntry` only for errors.

File: exp/tools/horizon-demo/reporter.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+// NewLoggingStateReporter constructs a new LoggingStateReporter instance
+func NewLoggingStateReporter(logInterval int) *LoggingStateReporter {
+	return &LoggingStateReporter{
+		logInterval: logInterval,
+	}
+}
+
+// OnStartState logs that the session has started reading from the history archive snapshot
+func (lr *LoggingStateReporter) OnStartState(sequence uint32) {
+	log.WithField("sequence", sequence).Info("Reading from History Archive Snapshot")
```
Comment (bartekn) [inline exp/tools/horizon-demo/reporter.go]: It looks like none of these print anything. It seems that default logger level is `WarnLevel`. Here's a code I used to fix this:
```go
// NewLoggingStateReporter constructs a new LoggingStateReporter instance
func NewLoggingStateReporter(logInterval int) *LoggingStateReporter {
	l := log.New()
	l.SetLevel(log.InfoLevel)

	return &LoggingStateReporter{
		logInterval: logInterval,
		log:         l,
	}
}
```

**Comment by tamirms** [comment]
@bartekn that is correct, if concurrency is removed then the output looks like:

```
API server listening at: 127.0.0.1:2345
INFO[2019-07-17T19:40:58.866+03:00] Reading from History Archive Snapshot         pid=42109 sequence=24877695
INFO[2019-07-17T19:41:05.085+03:00] Processed entries from History Archive Snapshot  numEntries=50000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:08.284+03:00] Processed entries from History Archive Snapshot  numEntries=100000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:10.265+03:00] Processed entries from History Archive Snapshot  numEntries=150000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:11.850+03:00] Processed entries from History Archive Snapshot  numEntries=200000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:12.964+03:00] Processed entries from History Archive Snapshot  numEntries=250000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:14.021+03:00] Processed entries from History Archive Snapshot  numEntries=300000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:15.110+03:00] Processed entries from History Archive Snapshot  numEntries=350000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:16.214+03:00] Processed entries from History Archive Snapshot  numEntries=400000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:18.024+03:00] Processed entries from History Archive Snapshot  numEntries=450000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:19.994+03:00] Processed entries from History Archive Snapshot  numEntries=500000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:21.266+03:00] Processed entries from History Archive Snapshot  numEntries=550000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:25.323+03:00] Processed entries from History Archive Snapshot  numEntries=600000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:30.838+03:00] Processed entries from History Archive Snapshot  numEntries=650000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:32.603+03:00] Processed entries from History Archive Snapshot  numEntries=700000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:33.689+03:00] Processed entries from History Archive Snapshot  numEntries=750000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:34.813+03:00] Processed entries from History Archive Snapshot  numEntries=800000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:36.463+03:00] Processed entries from History Archive Snapshot  numEntries=850000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:40.017+03:00] Processed entries from History Archive Snapshot  numEntries=900000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:43.556+03:00] Processed entries from History Archive Snapshot  numEntries=950000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:47.327+03:00] Processed entries from History Archive Snapshot  numEntries=1000000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:49.681+03:00] Processed entries from History Archive Snapshot  numEntries=1050000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:50.703+03:00] Processed entries from History Archive Snapshot  numEntries=1100000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:51.713+03:00] Processed entries from History Archive Snapshot  numEntries=1150000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:52.764+03:00] Processed entries from History Archive Snapshot  numEntries=1200000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:53.826+03:00] Processed entries from History Archive Snapshot  numEntries=1250000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:54.846+03:00] Processed entries from History Archive Snapshot  numEntries=1300000 pid=42109 sequence=24877695
INFO[2019-07-17T19:41:57.625+03:00] Processed entries from History Archive Snapshot  numEntries=1350000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:02.759+03:00] Processed entries from History Archive Snapshot  numEntries=1400000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:08.063+03:00] Processed entries from History Archive Snapshot  numEntries=1450000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:10.014+03:00] Processed entries from History Archive Snapshot  numEntries=1500000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:11.770+03:00] Processed entries from History Archive Snapshot  numEntries=1550000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:13.292+03:00] Processed entries from History Archive Snapshot  numEntries=1600000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:14.911+03:00] Processed entries from History Archive Snapshot  numEntries=1650000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:16.521+03:00] Processed entries from History Archive Snapshot  numEntries=1700000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:18.083+03:00] Processed entries from History Archive Snapshot  numEntries=1750000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:19.663+03:00] Processed entries from History Archive Snapshot  numEntries=1800000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:21.265+03:00] Processed entries from History Archive Snapshot  numEntries=1850000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:22.817+03:00] Processed entries from History Archive Snapshot  numEntries=1900000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:25.917+03:00] Processed entries from History Archive Snapshot  numEntries=1950000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:30.305+03:00] Processed entries from History Archive Snapshot  numEntries=2000000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:34.803+03:00] Processed entries from History Archive Snapshot  numEntries=2050000 pid=42109 sequence=24877695
INFO[2019-07-17T19:42:39.762+03:00] Processed entries from History Archive Snapshot  numEntries=2100000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:44.271+03:00] Processed entries from History Archive Snapshot  numEntries=2150000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:45.139+03:00] Processed entries from History Archive Snapshot  numEntries=2200000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:45.979+03:00] Processed entries from History Archive Snapshot  numEntries=2250000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:46.836+03:00] Processed entries from History Archive Snapshot  numEntries=2300000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:47.719+03:00] Processed entries from History Archive Snapshot  numEntries=2350000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:48.601+03:00] Processed entries from History Archive Snapshot  numEntries=2400000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:49.482+03:00] Processed entries from History Archive Snapshot  numEntries=2450000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:50.324+03:00] Processed entries from History Archive Snapshot  numEntries=2500000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:51.203+03:00] Processed entries from History Archive Snapshot  numEntries=2550000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:52.098+03:00] Processed entries from History Archive Snapshot  numEntries=2600000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:53.291+03:00] Processed entries from History Archive Snapshot  numEntries=2650000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:54.146+03:00] Processed entries from History Archive Snapshot  numEntries=2700000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:54.994+03:00] Processed entries from History Archive Snapshot  numEntries=2750000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:55.852+03:00] Processed entries from History Archive Snapshot  numEntries=2800000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:56.777+03:00] Processed entries from History Archive Snapshot  numEntries=2850000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:57.726+03:00] Processed entries from History Archive Snapshot  numEntries=2900000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:58.623+03:00] Processed entries from History Archive Snapshot  numEntries=2950000 pid=42109 sequence=24877695
INFO[2019-07-17T19:43:59.510+03:00] Processed entries from History Archive Snapshot  numEntries=3000000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:00.397+03:00] Processed entries from History Archive Snapshot  numEntries=3050000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:01.349+03:00] Processed entries from History Archive Snapshot  numEntries=3100000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:02.269+03:00] Processed entries from History Archive Snapshot  numEntries=3150000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:03.563+03:00] Processed entries from History Archive Snapshot  numEntries=3200000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:04.473+03:00] Processed entries from History Archive Snapshot  numEntries=3250000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:05.376+03:00] Processed entries from History Archive Snapshot  numEntries=3300000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:06.297+03:00] Processed entries from History Archive Snapshot  numEntries=3350000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:07.182+03:00] Processed entries from History Archive Snapshot  numEntries=3400000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:08.090+03:00] Processed entries from History Archive Snapshot  numEntries=3450000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:08.993+03:00] Processed entries from History Archive Snapshot  numEntries=3500000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:09.897+03:00] Processed entries from History Archive Snapshot  numEntries=3550000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:10.813+03:00] Processed entries from History Archive Snapshot  numEntries=3600000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:13.518+03:00] Processed entries from History Archive Snapshot  numEntries=3650000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:15.456+03:00] Processed entries from History Archive Snapshot  numEntries=3700000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:17.201+03:00] Processed entries from History Archive Snapshot  numEntries=3750000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:18.436+03:00] Processed entries from History Archive Snapshot  numEntries=3800000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:19.652+03:00] Processed entries from History Archive Snapshot  numEntries=3850000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:20.771+03:00] Processed entries from History Archive Snapshot  numEntries=3900000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:21.705+03:00] Processed entries from History Archive Snapshot  numEntries=3950000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:22.599+03:00] Processed entries from History Archive Snapshot  numEntries=4000000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:23.491+03:00] Processed entries from History Archive Snapshot  numEntries=4050000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:24.384+03:00] Processed entries from History Archive Snapshot  numEntries=4100000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:25.284+03:00] Processed entries from History Archive Snapshot  numEntries=4150000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:26.155+03:00] Processed entries from History Archive Snapshot  numEntries=4200000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:27.091+03:00] Processed entries from History Archive Snapshot  numEntries=4250000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:27.988+03:00] Processed entries from History Archive Snapshot  numEntries=4300000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:29.311+03:00] Processed entries from History Archive Snapshot  numEntries=4350000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:30.187+03:00] Processed entries from History Archive Snapshot  numEntries=4400000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:31.060+03:00] Processed entries from History Archive Snapshot  numEntries=4450000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:32.259+03:00] Processed entries from History Archive Snapshot  numEntries=4500000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:33.741+03:00] Processed entries from History Archive Snapshot  numEntries=4550000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:35.703+03:00] Processed entries from History Archive Snapshot  numEntries=4600000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:37.121+03:00] Processed entries from History Archive Snapshot  numEntries=4650000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:38.538+03:00] Processed entries from History Archive Snapshot  numEntries=4700000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:39.979+03:00] Processed entries from History Archive Snapshot  numEntries=4750000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:41.378+03:00] Processed entries from History Archive Snapshot  numEntries=4800000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:42.780+03:00] Processed entries from History Archive Snapshot  numEntries=4850000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:44.173+03:00] Processed entries from History Archive Snapshot  numEntries=4900000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:45.592+03:00] Processed entries from History Archive Snapshot  numEntries=4950000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:47.124+03:00] Processed entries from History Archive Snapshot  numEntries=5000000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:48.587+03:00] Processed entries from History Archive Snapshot  numEntries=5050000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:50.031+03:00] Processed entries from History Archive Snapshot  numEntries=5100000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:51.477+03:00] Processed entries from History Archive Snapshot  numEntries=5150000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:53.417+03:00] Processed entries from History Archive Snapshot  numEntries=5200000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:54.900+03:00] Processed entries from History Archive Snapshot  numEntries=5250000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:56.474+03:00] Processed entries from History Archive Snapshot  numEntries=5300000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:58.079+03:00] Processed entries from History Archive Snapshot  numEntries=5350000 pid=42109 sequence=24877695
INFO[2019-07-17T19:44:59.611+03:00] Processed entries from History Archive Snapshot  numEntries=5400000 pid=42109 sequence=24877695
INFO[2019-07-17T19:45:01.065+03:00] Processed entries from History Archive Snapshot  numEntries=5450000 pid=42109 sequence=24877695
INFO[2019-07-17T19:45:02.510+03:00] Processed entries from History Archive Snapshot  numEntries=5500000 pid=42109 sequence=24877695
INFO[2019-07-17T19:45:03.971+03:00] Processed entries from History Archive Snapshot  numEntries=5550000 pid=42109 sequence=24877695
INFO[2019-07-17T19:45:05.374+03:00] Processed entries from History Archive Snapshot  numEntries=5600000 pid=42109 sequence=24877695
INFO[2019-07-17T19:45:05.613+03:00] Finished processing History Archive Snapshot  elapsedSeconds=246.743631046 error="<nil>" numEntries=5612340 pid=42109 sequence=24877695 shutdown=false
```

**Comment by tamirms** [comment]
@bartekn can you take another look? i've pushed some fixes to the branch

**Pull request merged**